python/cudf/cudf/core/column/categorical.py

```diff
@@ -967,26 +973,42 @@ def find_and_replace(self, to_replace, replacement, all_nan):
         """
         Return col with *to_replace* replaced with *replacement*.
         """
-        replaced = column.as_column(self.cat().codes)
 
-        to_replace_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in to_replace], dtype=replaced.dtype
-            )
+        old_cats = cudf.DataFrame()
+        old_cats["cats"] = column.as_column(self.dtype.categories)
+        new_cats = old_cats.copy(deep=True)
+
+        old_cats["cats_replace"] = old_cats["cats"].replace(
+            to_replace, replacement
         )
-        replacement_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in replacement],
-                dtype=replaced.dtype,
-            )
+
+        dtype_replace = cudf.Series(replacement)
+        dtype_replace[dtype_replace.isin(old_cats["cats"])] = None
+        new_cats["cats"] = new_cats["cats"].replace(to_replace, dtype_replace)
+
+        bmask = new_cats["cats"]._column.notna()
+        new_cats = cudf.DataFrame(
+            {"cats": new_cats["cats"]._column.apply_boolean_mask(bmask)}
+        ).reset_index()
+
+        catmap = old_cats.merge(
+            new_cats, left_on="cats_replace", right_on="cats", how="inner"
+        )
+
+        to_replace_col = column.as_column(catmap.index).astype(
+            self.cat().codes.dtype
+        )
+        replacement_col = catmap["index"]._column.astype(
+            self.cat().codes.dtype
         )
 
```

Comment:
I found it difficult to understand the logic being implemented in this function - could we add a few comments describing how we're handling replacing categoricals?

Reply:
Agreed. The case that I had a hard time handling (hence all the logic) is one where we want to replace a category with a different one that already exists as part of the set. Originally I had hoped I could just replace the underlying ints with the int that corresponds to the 'destination' category. But that can leave a hole in the set of ints.

A concrete example is if we had categories `['a', 'b', 'c', 'd']` represented by the underlying ints `[0, 1, 2, 3]` and we want to replace `'b'` with `'d'`. If we just replaced all the `1`'s with `3`'s in the child column, we'd then have the ints `[0,2,3]` representing the categories. I think we actually support this fine, but pandas doesn't. So things like `repr` and `to_pandas` break.

Reply:
Hmm. Sounds like something that should be tackled at the `to_pandas` level then, so that we don't have to complicate the logic everywhere else. Thoughts?

Reply:
Not suggesting we do that in this PR though :)

Reply:
I considered that for a while. The way I pictured that having to work is we'd have to scan and detect the situation at the top of the `to_pandas` function and then possibly apply the same logic if conditions are fulfilled. I figured that meant extra work every time we ever call `to_pandas` on what might already be a valid categorical column, as opposed to just fixing it in this edge case where we can reasonably expect it to be wrong. I'm not sure there's any other cases where this could happen, but I wouldn't be against separating this code out into some kind of reusable `fix_categories` function so that if this does come up again no one needs to reinvent the wheel.

Reply:
Understood - let's leave as is for now, and if we come across this situation again (working around the limitation that Pandas doesn't handle codes like `[0, 2, 3]`), we can think about building a utility.

Reply:
This seems to work fine for me:
```
In [15]: pd.Series(pd.Categorical.from_codes([1, 0, 1, 3], ['a', 'b', 'c', 'd']))
Out[15]:
0    b
1    a
2    b
3    d
dtype: category
Categories (4, object): [a, b, c, d]
```

Reply:
That situation is technically valid because you still have four categories. But now suppose you wanted to replace `b` with `d`. You'd then be constructing the equivalent of this situation:

`pd.Series(pd.Categorical.from_codes([3,0,3,3], ['a','c','d]))`

Which will yield 
```
ValueError: codes need to be between -1 and len(categories)-1
```

Reply:
Yea you're out of bounds then, you'd need to replace only the values, not the dictionary. So the dictionary should still be `['a', 'b', 'c', 'd']`

Reply:
That is exactly what pandas 0.25.3 did. This however is the pandas 1.0 behavior, and what served as the impetus for reworking this:

```
>>> x = pd.Series(pd.Categorical.from_codes([1, 0, 1, 3], ['a', 'b', 'c', 'd']))
>>> x.replace('b','d')
0    d
1    a
2    d
3    d
dtype: category
Categories (3, object): [a, c, d]
```

Reply:
Yea, so sounds like we just need a comment explaining / surrounding this explanation of behavior.